// @cardano-sdk/core/dist/esm/Asset/NftMetadata/fromMetadatum.js
import { AssetId, AssetName } from '../../Cardano/types/Asset.js';
import { ImageMediaType, MediaType, Uri } from './types.js';
import { InvalidFileError } from './errors.js';
import { asMetadatumArray, asMetadatumMap } from '../../util/metadatum.js';
import { asString } from './util.js';
import { isNotNil } from '@cardano-sdk/util';
import difference from 'lodash/difference.js';
const isString = (obj) => typeof obj === 'string';
const VersionRegExp = /^\d+\.?\d?$/;
const metadatumToString = (metadatum) => {
    if (Array.isArray(metadatum)) {
        const result = metadatum.map(asString);
        if (result.some((str) => typeof str === 'undefined')) {
            return undefined;
        }
        return result.join('');
    }
    return asString(metadatum);
};
const mapOtherProperties = (metadata, primaryProperties) => {
    const extraProperties = difference([...metadata.keys()].filter(isString), primaryProperties);
    if (extraProperties.length === 0)
        return;
    return extraProperties.reduce((result, key) => {
        result.set(key, metadata.get(key));
        return result;
    }, new Map());
};
const missingFieldLogMessage = (fieldType, assetId, rootLevel) => `Omitting cip25 ${rootLevel ? 'root' : 'file'} metadata: missing "${fieldType}". AssetId: ${assetId}`;
const mapFile = (metadatum, assetId, logger) => {
    const file = asMetadatumMap(metadatum);
    if (!file)
        throw new InvalidFileError();
    const name = asString(file.get('name'));
    const mediaType = asString(file.get('mediaType'));
    if (!mediaType) {
        logger.warn(missingFieldLogMessage('mediaType', assetId, false));
        return null;
    }
    const src = metadatumToString(file.get('src'));
    if (!src) {
        logger.warn(missingFieldLogMessage('source', assetId, false));
        return null;
    }
    return {
        mediaType: MediaType(mediaType),
        name,
        otherProperties: mapOtherProperties(file, ['mediaType', 'name', 'src']),
        src: Uri(src)
    };
};
const getPolicyMetadata = (policy, policyId) => asMetadatumMap(policy.get(policyId) ||
    (() => {
        for (const [key, value] of policy.entries()) {
            if (ArrayBuffer.isView(key) && Buffer.from(key).toString('hex') === policyId)
                return value;
        }
    })());
const getAssetMetadata = (policy, assetName) => asMetadatumMap(policy.get(assetName) ||
    policy.get(Buffer.from(assetName, 'hex').toString('utf8')) ||
    (() => {
        for (const [key, value] of policy.entries()) {
            if (ArrayBuffer.isView(key) && Buffer.from(key).toString('hex') === assetName)
                return value;
        }
    })());
const getName = (assetMetadata, version, asset, logger, stripInvisibleCharacters = false) => {
    const name = asString(assetMetadata.get('name'));
    if (name)
        return name;
    if (version === '1.0') {
        try {
            return AssetName.toUTF8(asset.name, stripInvisibleCharacters);
        }
        catch (error) {
            logger.warn(error);
        }
    }
};
const parseVersion = (version) => {
    if (!version)
        return '1.0';
    if (typeof version === 'bigint') {
        return `${version}.0`;
    }
    const stringVersion = asString(version);
    if (stringVersion && VersionRegExp.test(stringVersion)) {
        return `${Number(stringVersion)}.0`;
    }
};
export const fromMetadatum = (asset, metadata, logger, strict = false) => {
    const cip25Metadata = metadata?.get(721n);
    if (!cip25Metadata)
        return null;
    const cip25MetadatumMap = asMetadatumMap(cip25Metadata);
    if (!cip25MetadatumMap)
        return null;
    const policy = getPolicyMetadata(cip25MetadatumMap, asset.policyId);
    if (!policy)
        return null;
    const version = parseVersion(policy.get('version'));
    if (!version)
        return null;
    const assetMetadata = getAssetMetadata(policy, asset.name);
    if (!assetMetadata)
        return null;
    const name = getName(assetMetadata, version, asset, logger, true);
    const image = metadatumToString(assetMetadata.get('image'));
    const assetId = AssetId.fromParts(asset.policyId, asset.name);
    if ((strict && !name) || !image) {
        logger.warn(missingFieldLogMessage(!name ? 'name' : 'image', assetId, true));
        return null;
    }
    const mediaType = asString(assetMetadata.get('mediaType'));
    const files = asMetadatumArray(assetMetadata.get('files'));
    try {
        return {
            description: metadatumToString(assetMetadata.get('description')),
            files: files?.map((file) => mapFile(file, assetId, logger)).filter(isNotNil),
            image: Uri(image),
            mediaType: mediaType ? ImageMediaType(mediaType) : undefined,
            name: name || '',
            otherProperties: mapOtherProperties(assetMetadata, ['name', 'image', 'mediaType', 'description', 'files']),
            version
        };
    }
    catch (error) {
        logger.warn('Invalid CIP-25 metadata', assetMetadata, error);
        return null;
    }
};
