// @cardano-sdk/core/dist/cjs/Provider/StakePoolProvider/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SortFields = exports.isPoolROSSortField = exports.isPoolAPYSortField = exports.isPoolMetricsSortField = exports.isPoolDataSortField = exports.PoolROSSortFields = exports.PoolAPYSortFields = exports.PoolMetricsSortFields = exports.PoolDataSortFields = void 0;
exports.PoolDataSortFields = ['cost', 'name', 'margin', 'pledge', 'ticker'];
exports.PoolMetricsSortFields = ['blocks', 'liveStake', 'saturation'];
exports.PoolAPYSortFields = ['apy'];
exports.PoolROSSortFields = ['ros', 'lastRos'];
const isPoolDataSortField = (value) => exports.PoolDataSortFields.includes(value);
exports.isPoolDataSortField = isPoolDataSortField;
const isPoolMetricsSortField = (value) => exports.PoolMetricsSortFields.includes(value);
exports.isPoolMetricsSortField = isPoolMetricsSortField;
const isPoolAPYSortField = (value) => exports.PoolAPYSortFields.includes(value);
exports.isPoolAPYSortField = isPoolAPYSortField;
const isPoolROSSortField = (value) => exports.PoolROSSortFields.includes(value);
exports.isPoolROSSortField = isPoolROSSortField;
exports.SortFields = [...exports.PoolDataSortFields, ...exports.PoolMetricsSortFields, ...exports.PoolAPYSortFields, ...exports.PoolROSSortFields];
