// @cardano-sdk/crypto/dist/cjs/Ed25519e/Ed25519KeyHash.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Ed25519KeyHash_hash;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ed25519KeyHash = exports.ED25519_PUBLIC_KEY_HASH_LENGTH = void 0;
const hexTypes_1 = require("../hexTypes");
const util_1 = require("@cardano-sdk/util");
exports.ED25519_PUBLIC_KEY_HASH_LENGTH = 28;
class Ed25519KeyHash {
    constructor(hash) {
        _Ed25519KeyHash_hash.set(this, void 0);
        __classPrivateFieldSet(this, _Ed25519KeyHash_hash, hash, "f");
    }
    static fromBytes(hash) {
        if (hash.length !== exports.ED25519_PUBLIC_KEY_HASH_LENGTH)
            throw new util_1.InvalidArgumentError('hash', `Hash should be ${exports.ED25519_PUBLIC_KEY_HASH_LENGTH} bytes; however ${hash.length} bytes were provided.`);
        return new Ed25519KeyHash(hash);
    }
    static fromHex(hash) {
        return Ed25519KeyHash.fromBytes(Buffer.from(hash, 'hex'));
    }
    bytes() {
        return __classPrivateFieldGet(this, _Ed25519KeyHash_hash, "f");
    }
    hex() {
        return (0, hexTypes_1.Ed25519KeyHashHex)(Buffer.from(__classPrivateFieldGet(this, _Ed25519KeyHash_hash, "f")).toString('hex'));
    }
}
exports.Ed25519KeyHash = Ed25519KeyHash;
_Ed25519KeyHash_hash = new WeakMap();
