// @cardano-sdk/crypto/dist/cjs/Ed25519e/Ed25519Signature.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Ed25519Signature_signature;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ed25519Signature = exports.ED25519_SIGNATURE_LENGTH = void 0;
const hexTypes_1 = require("../hexTypes");
const util_1 = require("@cardano-sdk/util");
exports.ED25519_SIGNATURE_LENGTH = 64;
class Ed25519Signature {
    constructor(signature) {
        _Ed25519Signature_signature.set(this, void 0);
        __classPrivateFieldSet(this, _Ed25519Signature_signature, signature, "f");
    }
    static fromBytes(signature) {
        if (signature.length !== exports.ED25519_SIGNATURE_LENGTH)
            throw new util_1.InvalidArgumentError('signature', `signature should be ${exports.ED25519_SIGNATURE_LENGTH} bytes; however ${signature.length} bytes were provided.`);
        return new Ed25519Signature(signature);
    }
    static fromHex(signature) {
        return Ed25519Signature.fromBytes(Buffer.from(signature, 'hex'));
    }
    bytes() {
        return __classPrivateFieldGet(this, _Ed25519Signature_signature, "f");
    }
    hex() {
        return (0, hexTypes_1.Ed25519SignatureHex)(Buffer.from(__classPrivateFieldGet(this, _Ed25519Signature_signature, "f")).toString('hex'));
    }
}
exports.Ed25519Signature = Ed25519Signature;
_Ed25519Signature_signature = new WeakMap();
