// @cardano-sdk/core/dist/cjs/Serialization/Certificates/MoveInstantaneousReward/MoveInstantaneousReward.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _MoveInstantaneousReward_toOtherPot, _MoveInstantaneousReward_toStakeCreds, _MoveInstantaneousReward_kind, _MoveInstantaneousReward_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MoveInstantaneousReward = void 0;
const CBOR_1 = require("../../CBOR");
const CertificateKind_1 = require("../CertificateKind");
const util_1 = require("@cardano-sdk/util");
const Certificate_1 = require("../../../Cardano/types/Certificate");
const MoveInstantaneousRewardToOtherPot_1 = require("./MoveInstantaneousRewardToOtherPot");
const MoveInstantaneousRewardToStakeCreds_1 = require("./MoveInstantaneousRewardToStakeCreds");
const EMBEDDED_GROUP_SIZE = 2;
class MoveInstantaneousReward {
    constructor() {
        _MoveInstantaneousReward_toOtherPot.set(this, void 0);
        _MoveInstantaneousReward_toStakeCreds.set(this, void 0);
        _MoveInstantaneousReward_kind.set(this, void 0);
        _MoveInstantaneousReward_originalBytes.set(this, undefined);
    }
    toCbor() {
        if (__classPrivateFieldGet(this, _MoveInstantaneousReward_originalBytes, "f"))
            return __classPrivateFieldGet(this, _MoveInstantaneousReward_originalBytes, "f");
        let cbor;
        switch (__classPrivateFieldGet(this, _MoveInstantaneousReward_kind, "f")) {
            case Certificate_1.MirCertificateKind.ToOtherPot:
                cbor = __classPrivateFieldGet(this, _MoveInstantaneousReward_toOtherPot, "f").toCbor();
                break;
            case Certificate_1.MirCertificateKind.ToStakeCreds:
                cbor = __classPrivateFieldGet(this, _MoveInstantaneousReward_toStakeCreds, "f").toCbor();
                break;
            default:
                throw new util_1.InvalidStateError(`Unexpected kind value: ${__classPrivateFieldGet(this, _MoveInstantaneousReward_kind, "f")}`);
        }
        const writer = new CBOR_1.CborWriter();
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(CertificateKind_1.CertificateKind.MoveInstantaneousRewards);
        writer.writeEncodedValue(Buffer.from(cbor, 'hex'));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        let elementsCount = reader.readStartArray();
        if (elementsCount !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected elements size ${EMBEDDED_GROUP_SIZE}, but got ${elementsCount}`);
        const kind = Number(reader.readInt());
        if (kind !== CertificateKind_1.CertificateKind.MoveInstantaneousRewards)
            throw new util_1.InvalidArgumentError('cbor', `Expected certificate kind ${CertificateKind_1.CertificateKind.MoveInstantaneousRewards}, but got ${kind}`);
        const embeddedCbor = util_1.HexBlob.fromBytes(reader.readEncodedValue());
        const embeddedCborReader = new CBOR_1.CborReader(embeddedCbor);
        elementsCount = embeddedCborReader.readStartArray();
        if (elementsCount !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected elements size ${EMBEDDED_GROUP_SIZE}, but got ${elementsCount}`);
        const cert = new MoveInstantaneousReward();
        embeddedCborReader.readInt();
        if (embeddedCborReader.peekState() === CBOR_1.CborReaderState.UnsignedInteger) {
            __classPrivateFieldSet(cert, _MoveInstantaneousReward_toOtherPot, MoveInstantaneousRewardToOtherPot_1.MoveInstantaneousRewardToOtherPot.fromCbor(embeddedCbor), "f");
            __classPrivateFieldSet(cert, _MoveInstantaneousReward_kind, Certificate_1.MirCertificateKind.ToOtherPot, "f");
        }
        else if (embeddedCborReader.peekState() === CBOR_1.CborReaderState.StartArray ||
            embeddedCborReader.peekState() === CBOR_1.CborReaderState.StartMap) {
            __classPrivateFieldSet(cert, _MoveInstantaneousReward_toStakeCreds, MoveInstantaneousRewardToStakeCreds_1.MoveInstantaneousRewardToStakeCreds.fromCbor(embeddedCbor), "f");
            __classPrivateFieldSet(cert, _MoveInstantaneousReward_kind, Certificate_1.MirCertificateKind.ToStakeCreds, "f");
        }
        else {
            throw new util_1.InvalidArgumentError('cbor', 'Invalid CBOR string');
        }
        __classPrivateFieldSet(cert, _MoveInstantaneousReward_originalBytes, cbor, "f");
        return cert;
    }
    toCore() {
        let core;
        switch (__classPrivateFieldGet(this, _MoveInstantaneousReward_kind, "f")) {
            case Certificate_1.MirCertificateKind.ToOtherPot:
                core = __classPrivateFieldGet(this, _MoveInstantaneousReward_toOtherPot, "f").toCore();
                break;
            case Certificate_1.MirCertificateKind.ToStakeCreds:
                core = __classPrivateFieldGet(this, _MoveInstantaneousReward_toStakeCreds, "f").toCore();
                break;
            default:
                throw new util_1.InvalidStateError(`Unexpected kind value: ${__classPrivateFieldGet(this, _MoveInstantaneousReward_kind, "f")}`);
        }
        return core;
    }
    static fromCore(cert) {
        const mirCert = new MoveInstantaneousReward();
        switch (cert.kind) {
            case Certificate_1.MirCertificateKind.ToOtherPot:
                __classPrivateFieldSet(mirCert, _MoveInstantaneousReward_toOtherPot, MoveInstantaneousRewardToOtherPot_1.MoveInstantaneousRewardToOtherPot.fromCore(cert), "f");
                __classPrivateFieldSet(mirCert, _MoveInstantaneousReward_kind, Certificate_1.MirCertificateKind.ToOtherPot, "f");
                break;
            case Certificate_1.MirCertificateKind.ToStakeCreds:
                __classPrivateFieldSet(mirCert, _MoveInstantaneousReward_toStakeCreds, MoveInstantaneousRewardToStakeCreds_1.MoveInstantaneousRewardToStakeCreds.fromCore(cert), "f");
                __classPrivateFieldSet(mirCert, _MoveInstantaneousReward_kind, Certificate_1.MirCertificateKind.ToStakeCreds, "f");
                break;
            default:
                throw new util_1.InvalidStateError(`Unexpected kind value: ${cert.kind}`);
        }
        return mirCert;
    }
    static newToOtherPot(mirCert) {
        const cert = new MoveInstantaneousReward();
        __classPrivateFieldSet(cert, _MoveInstantaneousReward_toOtherPot, mirCert, "f");
        __classPrivateFieldSet(cert, _MoveInstantaneousReward_kind, Certificate_1.MirCertificateKind.ToOtherPot, "f");
        return cert;
    }
    static newToStakeCreds(mirCert) {
        const cert = new MoveInstantaneousReward();
        __classPrivateFieldSet(cert, _MoveInstantaneousReward_toStakeCreds, mirCert, "f");
        __classPrivateFieldSet(cert, _MoveInstantaneousReward_kind, Certificate_1.MirCertificateKind.ToStakeCreds, "f");
        return cert;
    }
    kind() {
        return __classPrivateFieldGet(this, _MoveInstantaneousReward_kind, "f");
    }
    asToOtherPot() {
        return __classPrivateFieldGet(this, _MoveInstantaneousReward_toOtherPot, "f");
    }
    asToStakeCreds() {
        return __classPrivateFieldGet(this, _MoveInstantaneousReward_toStakeCreds, "f");
    }
}
exports.MoveInstantaneousReward = MoveInstantaneousReward;
_MoveInstantaneousReward_toOtherPot = new WeakMap(), _MoveInstantaneousReward_toStakeCreds = new WeakMap(), _MoveInstantaneousReward_kind = new WeakMap(), _MoveInstantaneousReward_originalBytes = new WeakMap();
