// @cardano-sdk/core/dist/cjs/Serialization/Certificates/GenesisKeyDelegation.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _GenesisKeyDelegation_genesisHash, _GenesisKeyDelegation_genesisDelegateHash, _GenesisKeyDelegation_vrfKeyHash, _GenesisKeyDelegation_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisKeyDelegation = void 0;
const Crypto = __importStar(require("@cardano-sdk/crypto"));
const CBOR_1 = require("../CBOR");
const CertificateKind_1 = require("./CertificateKind");
const Certificate_1 = require("../../Cardano/types/Certificate");
const util_1 = require("@cardano-sdk/util");
const EMBEDDED_GROUP_SIZE = 4;
class GenesisKeyDelegation {
    constructor(genesisHash, genesisDelegateHash, vrfKeyHash) {
        _GenesisKeyDelegation_genesisHash.set(this, void 0);
        _GenesisKeyDelegation_genesisDelegateHash.set(this, void 0);
        _GenesisKeyDelegation_vrfKeyHash.set(this, void 0);
        _GenesisKeyDelegation_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _GenesisKeyDelegation_genesisHash, genesisHash, "f");
        __classPrivateFieldSet(this, _GenesisKeyDelegation_genesisDelegateHash, genesisDelegateHash, "f");
        __classPrivateFieldSet(this, _GenesisKeyDelegation_vrfKeyHash, vrfKeyHash, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _GenesisKeyDelegation_originalBytes, "f"))
            return __classPrivateFieldGet(this, _GenesisKeyDelegation_originalBytes, "f");
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(CertificateKind_1.CertificateKind.GenesisKeyDelegation);
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _GenesisKeyDelegation_genesisHash, "f"), 'hex'));
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _GenesisKeyDelegation_genesisDelegateHash, "f"), 'hex'));
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _GenesisKeyDelegation_vrfKeyHash, "f"), 'hex'));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== CertificateKind_1.CertificateKind.GenesisKeyDelegation)
            throw new util_1.InvalidArgumentError('cbor', `Expected certificate kind ${CertificateKind_1.CertificateKind.GenesisKeyDelegation}, but got ${kind}`);
        const genesisHash = Crypto.Hash28ByteBase16(util_1.HexBlob.fromBytes(reader.readByteString()));
        const genesisDelegateHash = Crypto.Hash28ByteBase16(util_1.HexBlob.fromBytes(reader.readByteString()));
        const vrfKeyHash = Crypto.Hash32ByteBase16(util_1.HexBlob.fromBytes(reader.readByteString()));
        reader.readEndArray();
        const cert = new GenesisKeyDelegation(genesisHash, genesisDelegateHash, vrfKeyHash);
        __classPrivateFieldSet(cert, _GenesisKeyDelegation_originalBytes, cbor, "f");
        return cert;
    }
    toCore() {
        return {
            __typename: Certificate_1.CertificateType.GenesisKeyDelegation,
            genesisDelegateHash: __classPrivateFieldGet(this, _GenesisKeyDelegation_genesisDelegateHash, "f"),
            genesisHash: __classPrivateFieldGet(this, _GenesisKeyDelegation_genesisHash, "f"),
            vrfKeyHash: __classPrivateFieldGet(this, _GenesisKeyDelegation_vrfKeyHash, "f")
        };
    }
    static fromCore(cert) {
        return new GenesisKeyDelegation(cert.genesisHash, cert.genesisDelegateHash, cert.vrfKeyHash);
    }
    genesisHash() {
        return __classPrivateFieldGet(this, _GenesisKeyDelegation_genesisHash, "f");
    }
    setGenesisHash(genesisHash) {
        __classPrivateFieldSet(this, _GenesisKeyDelegation_genesisHash, genesisHash, "f");
        __classPrivateFieldSet(this, _GenesisKeyDelegation_originalBytes, undefined, "f");
    }
    genesisDelegateHash() {
        return __classPrivateFieldGet(this, _GenesisKeyDelegation_genesisDelegateHash, "f");
    }
    setGenesisDelegateHash(genesisDelegateHash) {
        __classPrivateFieldSet(this, _GenesisKeyDelegation_genesisDelegateHash, genesisDelegateHash, "f");
        __classPrivateFieldSet(this, _GenesisKeyDelegation_originalBytes, undefined, "f");
    }
    vrfKeyHash() {
        return __classPrivateFieldGet(this, _GenesisKeyDelegation_vrfKeyHash, "f");
    }
    setVrfKeyHash(vrfKeyHash) {
        __classPrivateFieldSet(this, _GenesisKeyDelegation_vrfKeyHash, vrfKeyHash, "f");
        __classPrivateFieldSet(this, _GenesisKeyDelegation_originalBytes, undefined, "f");
    }
}
exports.GenesisKeyDelegation = GenesisKeyDelegation;
_GenesisKeyDelegation_genesisHash = new WeakMap(), _GenesisKeyDelegation_genesisDelegateHash = new WeakMap(), _GenesisKeyDelegation_vrfKeyHash = new WeakMap(), _GenesisKeyDelegation_originalBytes = new WeakMap();
