// @cardano-sdk/core/dist/cjs/Serialization/Certificates/PoolRegistration.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PoolRegistration_params, _PoolRegistration_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PoolRegistration = void 0;
const CBOR_1 = require("../CBOR");
const CertificateKind_1 = require("./CertificateKind");
const Certificate_1 = require("../../Cardano/types/Certificate");
const util_1 = require("@cardano-sdk/util");
const PoolParams_1 = require("./PoolParams");
const EMBEDDED_GROUP_SIZE = PoolParams_1.PoolParams.subgroupCount + 1;
class PoolRegistration {
    constructor(params) {
        _PoolRegistration_params.set(this, void 0);
        _PoolRegistration_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _PoolRegistration_params, params, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _PoolRegistration_originalBytes, "f"))
            return __classPrivateFieldGet(this, _PoolRegistration_originalBytes, "f");
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(CertificateKind_1.CertificateKind.PoolRegistration);
        return __classPrivateFieldGet(this, _PoolRegistration_params, "f").toFlattenedCbor(writer);
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== CertificateKind_1.CertificateKind.PoolRegistration)
            throw new util_1.InvalidArgumentError('cbor', `Expected certificate kind ${CertificateKind_1.CertificateKind.PoolRegistration}, but got ${kind}`);
        const params = PoolParams_1.PoolParams.fromFlattenedCbor(reader);
        reader.readEndArray();
        const cert = new PoolRegistration(params);
        __classPrivateFieldSet(cert, _PoolRegistration_originalBytes, cbor, "f");
        return cert;
    }
    toCore() {
        return {
            __typename: Certificate_1.CertificateType.PoolRegistration,
            poolParameters: __classPrivateFieldGet(this, _PoolRegistration_params, "f").toCore()
        };
    }
    static fromCore(cert) {
        return new PoolRegistration(PoolParams_1.PoolParams.fromCore(cert.poolParameters));
    }
    poolParameters() {
        return __classPrivateFieldGet(this, _PoolRegistration_params, "f");
    }
    setPoolParameters(parameters) {
        __classPrivateFieldSet(this, _PoolRegistration_params, parameters, "f");
        __classPrivateFieldSet(this, _PoolRegistration_originalBytes, undefined, "f");
    }
}
exports.PoolRegistration = PoolRegistration;
_PoolRegistration_params = new WeakMap(), _PoolRegistration_originalBytes = new WeakMap();
