// @cardano-sdk/core/dist/cjs/Serialization/Certificates/PoolRetirement.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PoolRetirement_poolKeyHash, _PoolRetirement_epoch, _PoolRetirement_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PoolRetirement = void 0;
const Crypto = __importStar(require("@cardano-sdk/crypto"));
const CBOR_1 = require("../CBOR");
const CertificateKind_1 = require("./CertificateKind");
const Certificate_1 = require("../../Cardano/types/Certificate");
const Block_1 = require("../../Cardano/types/Block");
const util_1 = require("@cardano-sdk/util");
const StakePool_1 = require("../../Cardano/types/StakePool");
const EMBEDDED_GROUP_SIZE = 3;
class PoolRetirement {
    constructor(poolKeyHash, epoch) {
        _PoolRetirement_poolKeyHash.set(this, void 0);
        _PoolRetirement_epoch.set(this, void 0);
        _PoolRetirement_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _PoolRetirement_poolKeyHash, poolKeyHash, "f");
        __classPrivateFieldSet(this, _PoolRetirement_epoch, epoch, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _PoolRetirement_originalBytes, "f"))
            return __classPrivateFieldGet(this, _PoolRetirement_originalBytes, "f");
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(CertificateKind_1.CertificateKind.PoolRetirement);
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _PoolRetirement_poolKeyHash, "f"), 'hex'));
        writer.writeInt(__classPrivateFieldGet(this, _PoolRetirement_epoch, "f"));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== CertificateKind_1.CertificateKind.PoolRetirement)
            throw new util_1.InvalidArgumentError('cbor', `Expected certificate kind ${CertificateKind_1.CertificateKind.PoolRetirement}, but got ${kind}`);
        const poolKeyHash = Crypto.Ed25519KeyHashHex(util_1.HexBlob.fromBytes(reader.readByteString()));
        const epoch = reader.readInt();
        reader.readEndArray();
        const cert = new PoolRetirement(poolKeyHash, (0, Block_1.EpochNo)(Number(epoch)));
        __classPrivateFieldSet(cert, _PoolRetirement_originalBytes, cbor, "f");
        return cert;
    }
    toCore() {
        return {
            __typename: Certificate_1.CertificateType.PoolRetirement,
            epoch: __classPrivateFieldGet(this, _PoolRetirement_epoch, "f"),
            poolId: StakePool_1.PoolId.fromKeyHash(__classPrivateFieldGet(this, _PoolRetirement_poolKeyHash, "f"))
        };
    }
    static fromCore(cert) {
        return new PoolRetirement(StakePool_1.PoolId.toKeyHash(cert.poolId), cert.epoch);
    }
    poolKeyHash() {
        return __classPrivateFieldGet(this, _PoolRetirement_poolKeyHash, "f");
    }
    setPoolKeyHash(poolKeyHash) {
        __classPrivateFieldSet(this, _PoolRetirement_poolKeyHash, poolKeyHash, "f");
        __classPrivateFieldSet(this, _PoolRetirement_originalBytes, undefined, "f");
    }
    epoch() {
        return __classPrivateFieldGet(this, _PoolRetirement_epoch, "f");
    }
    setEpoch(epoch) {
        __classPrivateFieldSet(this, _PoolRetirement_epoch, epoch, "f");
        __classPrivateFieldSet(this, _PoolRetirement_originalBytes, undefined, "f");
    }
}
exports.PoolRetirement = PoolRetirement;
_PoolRetirement_poolKeyHash = new WeakMap(), _PoolRetirement_epoch = new WeakMap(), _PoolRetirement_originalBytes = new WeakMap();
