// @cardano-sdk/core/dist/cjs/Serialization/Certificates/RegisterDelegateRepresentative.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _RegisterDelegateRepresentative_drepCredential, _RegisterDelegateRepresentative_deposit, _RegisterDelegateRepresentative_anchor, _RegisterDelegateRepresentative_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegisterDelegateRepresentative = void 0;
const Crypto = __importStar(require("@cardano-sdk/crypto"));
const Common_1 = require("../Common");
const CBOR_1 = require("../CBOR");
const CertificateKind_1 = require("./CertificateKind");
const Certificate_1 = require("../../Cardano/types/Certificate");
const util_1 = require("@cardano-sdk/util");
const misc_1 = require("../../util/misc");
const EMBEDDED_GROUP_SIZE = 2;
class RegisterDelegateRepresentative {
    constructor(drepCredential, deposit, anchor) {
        _RegisterDelegateRepresentative_drepCredential.set(this, void 0);
        _RegisterDelegateRepresentative_deposit.set(this, void 0);
        _RegisterDelegateRepresentative_anchor.set(this, void 0);
        _RegisterDelegateRepresentative_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _RegisterDelegateRepresentative_drepCredential, drepCredential, "f");
        __classPrivateFieldSet(this, _RegisterDelegateRepresentative_deposit, deposit, "f");
        __classPrivateFieldSet(this, _RegisterDelegateRepresentative_anchor, anchor, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _RegisterDelegateRepresentative_originalBytes, "f"))
            return __classPrivateFieldGet(this, _RegisterDelegateRepresentative_originalBytes, "f");
        writer.writeStartArray(4);
        writer.writeInt(CertificateKind_1.CertificateKind.DrepRegistration);
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(__classPrivateFieldGet(this, _RegisterDelegateRepresentative_drepCredential, "f").type);
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _RegisterDelegateRepresentative_drepCredential, "f").hash, 'hex'));
        writer.writeInt(__classPrivateFieldGet(this, _RegisterDelegateRepresentative_deposit, "f"));
        if (__classPrivateFieldGet(this, _RegisterDelegateRepresentative_anchor, "f")) {
            writer.writeEncodedValue((0, misc_1.hexToBytes)(__classPrivateFieldGet(this, _RegisterDelegateRepresentative_anchor, "f").toCbor()));
        }
        else {
            writer.writeNull();
        }
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== 4)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of 4 elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== CertificateKind_1.CertificateKind.DrepRegistration)
            throw new util_1.InvalidArgumentError('cbor', `Expected certificate kind ${CertificateKind_1.CertificateKind.DrepRegistration}, but got ${kind}`);
        const credLength = reader.readStartArray();
        if (credLength !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const type = Number(reader.readInt());
        const hash = Crypto.Hash28ByteBase16(util_1.HexBlob.fromBytes(reader.readByteString()));
        reader.readEndArray();
        const deposit = reader.readInt();
        let anchor;
        if (reader.peekState() === CBOR_1.CborReaderState.Null) {
            reader.readNull();
        }
        else {
            anchor = Common_1.Anchor.fromCbor(util_1.HexBlob.fromBytes(reader.readEncodedValue()));
        }
        reader.readEndArray();
        const cert = new RegisterDelegateRepresentative({ hash, type }, deposit, anchor);
        __classPrivateFieldSet(cert, _RegisterDelegateRepresentative_originalBytes, cbor, "f");
        return cert;
    }
    toCore() {
        return {
            __typename: Certificate_1.CertificateType.RegisterDelegateRepresentative,
            anchor: __classPrivateFieldGet(this, _RegisterDelegateRepresentative_anchor, "f") ? __classPrivateFieldGet(this, _RegisterDelegateRepresentative_anchor, "f").toCore() : null,
            dRepCredential: __classPrivateFieldGet(this, _RegisterDelegateRepresentative_drepCredential, "f"),
            deposit: __classPrivateFieldGet(this, _RegisterDelegateRepresentative_deposit, "f")
        };
    }
    static fromCore(cert) {
        return new RegisterDelegateRepresentative(cert.dRepCredential, cert.deposit, cert.anchor ? Common_1.Anchor.fromCore(cert.anchor) : undefined);
    }
    credential() {
        return __classPrivateFieldGet(this, _RegisterDelegateRepresentative_drepCredential, "f");
    }
    deposit() {
        return __classPrivateFieldGet(this, _RegisterDelegateRepresentative_deposit, "f");
    }
    anchor() {
        return __classPrivateFieldGet(this, _RegisterDelegateRepresentative_anchor, "f");
    }
}
exports.RegisterDelegateRepresentative = RegisterDelegateRepresentative;
_RegisterDelegateRepresentative_drepCredential = new WeakMap(), _RegisterDelegateRepresentative_deposit = new WeakMap(), _RegisterDelegateRepresentative_anchor = new WeakMap(), _RegisterDelegateRepresentative_originalBytes = new WeakMap();
