// @cardano-sdk/core/dist/cjs/Serialization/Certificates/StakeVoteDelegation.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _StakeVoteDelegation_credential, _StakeVoteDelegation_poolKeyHash, _StakeVoteDelegation_dRep, _StakeVoteDelegation_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakeVoteDelegation = void 0;
const Crypto = __importStar(require("@cardano-sdk/crypto"));
const CBOR_1 = require("../CBOR");
const CertificateKind_1 = require("./CertificateKind");
const Certificate_1 = require("../../Cardano/types/Certificate");
const DRep_1 = require("./DRep");
const util_1 = require("@cardano-sdk/util");
const StakePool_1 = require("../../Cardano/types/StakePool");
const misc_1 = require("../../util/misc");
const EMBEDDED_GROUP_SIZE = 2;
class StakeVoteDelegation {
    constructor(stakeCredential, drep, poolKeyHash) {
        _StakeVoteDelegation_credential.set(this, void 0);
        _StakeVoteDelegation_poolKeyHash.set(this, void 0);
        _StakeVoteDelegation_dRep.set(this, void 0);
        _StakeVoteDelegation_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _StakeVoteDelegation_credential, stakeCredential, "f");
        __classPrivateFieldSet(this, _StakeVoteDelegation_dRep, drep, "f");
        __classPrivateFieldSet(this, _StakeVoteDelegation_poolKeyHash, poolKeyHash, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _StakeVoteDelegation_originalBytes, "f"))
            return __classPrivateFieldGet(this, _StakeVoteDelegation_originalBytes, "f");
        writer.writeStartArray(4);
        writer.writeInt(CertificateKind_1.CertificateKind.StakeVoteDelegation);
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(__classPrivateFieldGet(this, _StakeVoteDelegation_credential, "f").type);
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _StakeVoteDelegation_credential, "f").hash, 'hex'));
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _StakeVoteDelegation_poolKeyHash, "f"), 'hex'));
        writer.writeEncodedValue((0, misc_1.hexToBytes)(__classPrivateFieldGet(this, _StakeVoteDelegation_dRep, "f").toCbor()));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== 4)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of 4 elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== CertificateKind_1.CertificateKind.StakeVoteDelegation)
            throw new util_1.InvalidArgumentError('cbor', `Expected certificate kind ${CertificateKind_1.CertificateKind.StakeVoteDelegation}, but got ${kind}`);
        const credLength = reader.readStartArray();
        if (credLength !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const type = Number(reader.readInt());
        const hash = Crypto.Hash28ByteBase16(util_1.HexBlob.fromBytes(reader.readByteString()));
        reader.readEndArray();
        const poolKeyHash = Crypto.Ed25519KeyHashHex(util_1.HexBlob.fromBytes(reader.readByteString()));
        const dRep = DRep_1.DRep.fromCbor(util_1.HexBlob.fromBytes(reader.readEncodedValue()));
        reader.readEndArray();
        const cert = new StakeVoteDelegation({ hash, type }, dRep, poolKeyHash);
        __classPrivateFieldSet(cert, _StakeVoteDelegation_originalBytes, cbor, "f");
        return cert;
    }
    toCore() {
        return {
            __typename: Certificate_1.CertificateType.StakeVoteDelegation,
            dRep: __classPrivateFieldGet(this, _StakeVoteDelegation_dRep, "f").toCore(),
            poolId: StakePool_1.PoolId.fromKeyHash(__classPrivateFieldGet(this, _StakeVoteDelegation_poolKeyHash, "f")),
            stakeCredential: __classPrivateFieldGet(this, _StakeVoteDelegation_credential, "f")
        };
    }
    static fromCore(deleg) {
        return new StakeVoteDelegation(deleg.stakeCredential, DRep_1.DRep.fromCore(deleg.dRep), StakePool_1.PoolId.toKeyHash(deleg.poolId));
    }
    stakeCredential() {
        return __classPrivateFieldGet(this, _StakeVoteDelegation_credential, "f");
    }
    drep() {
        return __classPrivateFieldGet(this, _StakeVoteDelegation_dRep, "f");
    }
    poolKeyHash() {
        return __classPrivateFieldGet(this, _StakeVoteDelegation_poolKeyHash, "f");
    }
}
exports.StakeVoteDelegation = StakeVoteDelegation;
_StakeVoteDelegation_credential = new WeakMap(), _StakeVoteDelegation_poolKeyHash = new WeakMap(), _StakeVoteDelegation_dRep = new WeakMap(), _StakeVoteDelegation_originalBytes = new WeakMap();
