// @cardano-sdk/core/dist/cjs/Serialization/Certificates/StakeRegistrationDelegation.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _StakeRegistrationDelegation_credential, _StakeRegistrationDelegation_poolKeyHash, _StakeRegistrationDelegation_deposit, _StakeRegistrationDelegation_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakeRegistrationDelegation = void 0;
const Crypto = __importStar(require("@cardano-sdk/crypto"));
const CBOR_1 = require("../CBOR");
const CertificateKind_1 = require("./CertificateKind");
const Certificate_1 = require("../../Cardano/types/Certificate");
const util_1 = require("@cardano-sdk/util");
const StakePool_1 = require("../../Cardano/types/StakePool");
const EMBEDDED_GROUP_SIZE = 2;
class StakeRegistrationDelegation {
    constructor(stakeCredential, deposit, poolKeyHash) {
        _StakeRegistrationDelegation_credential.set(this, void 0);
        _StakeRegistrationDelegation_poolKeyHash.set(this, void 0);
        _StakeRegistrationDelegation_deposit.set(this, void 0);
        _StakeRegistrationDelegation_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _StakeRegistrationDelegation_credential, stakeCredential, "f");
        __classPrivateFieldSet(this, _StakeRegistrationDelegation_deposit, deposit, "f");
        __classPrivateFieldSet(this, _StakeRegistrationDelegation_poolKeyHash, poolKeyHash, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _StakeRegistrationDelegation_originalBytes, "f"))
            return __classPrivateFieldGet(this, _StakeRegistrationDelegation_originalBytes, "f");
        writer.writeStartArray(4);
        writer.writeInt(CertificateKind_1.CertificateKind.StakeRegistrationDelegation);
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(__classPrivateFieldGet(this, _StakeRegistrationDelegation_credential, "f").type);
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _StakeRegistrationDelegation_credential, "f").hash, 'hex'));
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _StakeRegistrationDelegation_poolKeyHash, "f"), 'hex'));
        writer.writeInt(__classPrivateFieldGet(this, _StakeRegistrationDelegation_deposit, "f"));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== 4)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of 4 elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== CertificateKind_1.CertificateKind.StakeRegistrationDelegation)
            throw new util_1.InvalidArgumentError('cbor', `Expected certificate kind ${CertificateKind_1.CertificateKind.StakeRegistrationDelegation}, but got ${kind}`);
        const credLength = reader.readStartArray();
        if (credLength !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const type = Number(reader.readInt());
        const hash = Crypto.Hash28ByteBase16(util_1.HexBlob.fromBytes(reader.readByteString()));
        reader.readEndArray();
        const poolKeyHash = Crypto.Ed25519KeyHashHex(util_1.HexBlob.fromBytes(reader.readByteString()));
        const deposit = reader.readInt();
        reader.readEndArray();
        const cert = new StakeRegistrationDelegation({ hash, type }, deposit, poolKeyHash);
        __classPrivateFieldSet(cert, _StakeRegistrationDelegation_originalBytes, cbor, "f");
        return cert;
    }
    toCore() {
        return {
            __typename: Certificate_1.CertificateType.StakeRegistrationDelegation,
            deposit: __classPrivateFieldGet(this, _StakeRegistrationDelegation_deposit, "f"),
            poolId: StakePool_1.PoolId.fromKeyHash(__classPrivateFieldGet(this, _StakeRegistrationDelegation_poolKeyHash, "f")),
            stakeCredential: __classPrivateFieldGet(this, _StakeRegistrationDelegation_credential, "f")
        };
    }
    static fromCore(deleg) {
        return new StakeRegistrationDelegation(deleg.stakeCredential, deleg.deposit, StakePool_1.PoolId.toKeyHash(deleg.poolId));
    }
    stakeCredential() {
        return __classPrivateFieldGet(this, _StakeRegistrationDelegation_credential, "f");
    }
    deposit() {
        return __classPrivateFieldGet(this, _StakeRegistrationDelegation_deposit, "f");
    }
    poolKeyHash() {
        return __classPrivateFieldGet(this, _StakeRegistrationDelegation_poolKeyHash, "f");
    }
}
exports.StakeRegistrationDelegation = StakeRegistrationDelegation;
_StakeRegistrationDelegation_credential = new WeakMap(), _StakeRegistrationDelegation_poolKeyHash = new WeakMap(), _StakeRegistrationDelegation_deposit = new WeakMap(), _StakeRegistrationDelegation_originalBytes = new WeakMap();
