// @cardano-sdk/core/dist/cjs/Serialization/Certificates/VoteDelegation.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _VoteDelegation_credential, _VoteDelegation_dRep, _VoteDelegation_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoteDelegation = void 0;
const Crypto = __importStar(require("@cardano-sdk/crypto"));
const CBOR_1 = require("../CBOR");
const CertificateKind_1 = require("./CertificateKind");
const Certificate_1 = require("../../Cardano/types/Certificate");
const DRep_1 = require("./DRep");
const util_1 = require("@cardano-sdk/util");
const misc_1 = require("../../util/misc");
const EMBEDDED_GROUP_SIZE = 2;
class VoteDelegation {
    constructor(stakeCredential, dRep) {
        _VoteDelegation_credential.set(this, void 0);
        _VoteDelegation_dRep.set(this, void 0);
        _VoteDelegation_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _VoteDelegation_credential, stakeCredential, "f");
        __classPrivateFieldSet(this, _VoteDelegation_dRep, dRep, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _VoteDelegation_originalBytes, "f"))
            return __classPrivateFieldGet(this, _VoteDelegation_originalBytes, "f");
        writer.writeStartArray(3);
        writer.writeInt(CertificateKind_1.CertificateKind.VoteDelegation);
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(__classPrivateFieldGet(this, _VoteDelegation_credential, "f").type);
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _VoteDelegation_credential, "f").hash, 'hex'));
        writer.writeEncodedValue((0, misc_1.hexToBytes)(__classPrivateFieldGet(this, _VoteDelegation_dRep, "f").toCbor()));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== 3)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of 3 elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== CertificateKind_1.CertificateKind.VoteDelegation)
            throw new util_1.InvalidArgumentError('cbor', `Expected certificate kind ${CertificateKind_1.CertificateKind.VoteDelegation}, but got ${kind}`);
        const credLength = reader.readStartArray();
        if (credLength !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const type = Number(reader.readInt());
        const hash = Crypto.Hash28ByteBase16(util_1.HexBlob.fromBytes(reader.readByteString()));
        reader.readEndArray();
        const dRep = DRep_1.DRep.fromCbor(util_1.HexBlob.fromBytes(reader.readEncodedValue()));
        reader.readEndArray();
        const cert = new VoteDelegation({ hash, type }, dRep);
        __classPrivateFieldSet(cert, _VoteDelegation_originalBytes, cbor, "f");
        return cert;
    }
    toCore() {
        return {
            __typename: Certificate_1.CertificateType.VoteDelegation,
            dRep: __classPrivateFieldGet(this, _VoteDelegation_dRep, "f").toCore(),
            stakeCredential: __classPrivateFieldGet(this, _VoteDelegation_credential, "f")
        };
    }
    static fromCore(deleg) {
        return new VoteDelegation(deleg.stakeCredential, DRep_1.DRep.fromCore(deleg.dRep));
    }
    stakeCredential() {
        return __classPrivateFieldGet(this, _VoteDelegation_credential, "f");
    }
    dRep() {
        return __classPrivateFieldGet(this, _VoteDelegation_dRep, "f");
    }
}
exports.VoteDelegation = VoteDelegation;
_VoteDelegation_credential = new WeakMap(), _VoteDelegation_dRep = new WeakMap(), _VoteDelegation_originalBytes = new WeakMap();
