// @cardano-sdk/crypto/dist/cjs/Bip32/Bip32PrivateKey.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _Bip32PrivateKey_key;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bip32PrivateKey = exports.BIP32_ED25519_PRIVATE_KEY_LENGTH = void 0;
const Bip32KeyDerivation = __importStar(require("./Bip32KeyDerivation"));
const hexTypes_1 = require("../hexTypes");
const Bip32PublicKey_1 = require("./Bip32PublicKey");
const Ed25519e_1 = require("../Ed25519e");
const util_1 = require("@cardano-sdk/util");
const pbkdf2_1 = require("pbkdf2");
const libsodium_wrappers_sumo_1 = __importDefault(require("libsodium-wrappers-sumo"));
const SCALAR_INDEX = 0;
const SCALAR_SIZE = 32;
const PBKDF2_ITERATIONS = 4096;
const PBKDF2_KEY_SIZE = 96;
const PBKDF2_DIGEST_ALGORITHM = 'sha512';
const CHAIN_CODE_INDEX = 64;
const CHAIN_CODE_SIZE = 32;
const clampScalar = (scalar) => {
    scalar[0] &= 248;
    scalar[31] &= 31;
    scalar[31] |= 64;
    return scalar;
};
const extendedScalar = (extendedKey) => extendedKey.slice(SCALAR_INDEX, SCALAR_SIZE);
exports.BIP32_ED25519_PRIVATE_KEY_LENGTH = 96;
class Bip32PrivateKey {
    constructor(key) {
        _Bip32PrivateKey_key.set(this, void 0);
        __classPrivateFieldSet(this, _Bip32PrivateKey_key, key, "f");
    }
    static fromBip39Entropy(entropy, password) {
        const xprv = (0, pbkdf2_1.pbkdf2Sync)(password, entropy, PBKDF2_ITERATIONS, PBKDF2_KEY_SIZE, PBKDF2_DIGEST_ALGORITHM);
        return Bip32PrivateKey.fromBytes(clampScalar(xprv));
    }
    static fromBytes(key) {
        if (key.length !== exports.BIP32_ED25519_PRIVATE_KEY_LENGTH)
            throw new util_1.InvalidArgumentError('key', `Key should be ${exports.BIP32_ED25519_PRIVATE_KEY_LENGTH} bytes; however ${key.length} bytes were provided.`);
        return new Bip32PrivateKey(key);
    }
    static fromHex(key) {
        return Bip32PrivateKey.fromBytes(Buffer.from(key, 'hex'));
    }
    async derive(derivationIndices) {
        await libsodium_wrappers_sumo_1.default.ready;
        let key = Buffer.from(__classPrivateFieldGet(this, _Bip32PrivateKey_key, "f"));
        for (const index of derivationIndices) {
            key = Bip32KeyDerivation.derivePrivate(key, index);
        }
        return Bip32PrivateKey.fromBytes(key);
    }
    toRawKey() {
        return Ed25519e_1.Ed25519PrivateKey.fromExtendedBytes(__classPrivateFieldGet(this, _Bip32PrivateKey_key, "f").slice(0, Ed25519e_1.EXTENDED_ED25519_PRIVATE_KEY_LENGTH));
    }
    async toPublic() {
        await libsodium_wrappers_sumo_1.default.ready;
        const scalar = extendedScalar(__classPrivateFieldGet(this, _Bip32PrivateKey_key, "f").slice(0, Ed25519e_1.EXTENDED_ED25519_PRIVATE_KEY_LENGTH));
        const publicKey = libsodium_wrappers_sumo_1.default.crypto_scalarmult_ed25519_base_noclamp(scalar);
        return Bip32PublicKey_1.Bip32PublicKey.fromBytes(Buffer.concat([publicKey, __classPrivateFieldGet(this, _Bip32PrivateKey_key, "f").slice(CHAIN_CODE_INDEX, CHAIN_CODE_INDEX + CHAIN_CODE_SIZE)]));
    }
    bytes() {
        return __classPrivateFieldGet(this, _Bip32PrivateKey_key, "f");
    }
    hex() {
        return (0, hexTypes_1.Bip32PrivateKeyHex)(Buffer.from(__classPrivateFieldGet(this, _Bip32PrivateKey_key, "f")).toString('hex'));
    }
}
exports.Bip32PrivateKey = Bip32PrivateKey;
_Bip32PrivateKey_key = new WeakMap();
