// @cardano-sdk/core/dist/cjs/util/time.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeSpan = exports.Days = exports.Hours = exports.Minutes = exports.Seconds = exports.Milliseconds = void 0;
const Milliseconds = (value) => value;
exports.Milliseconds = Milliseconds;
const Seconds = (value) => value;
exports.Seconds = Seconds;
const Minutes = (value) => value;
exports.Minutes = Minutes;
const Hours = (value) => value;
exports.Hours = Hours;
const Days = (value) => value;
exports.Days = Days;
exports.Hours.toDays = (value) => (value / 24);
exports.Days.toHours = (value) => (value * 24);
exports.Minutes.toHours = (value) => (value / 60);
exports.Hours.toMinutes = (value) => (value * 60);
exports.Seconds.toMinutes = (value) => (value / 60);
exports.Minutes.toSeconds = (value) => (value * 60);
exports.Seconds.toMilliseconds = (value) => (value * 1000);
exports.Milliseconds.toSeconds = (value) => (value / 1000);
class TimeSpan {
    constructor(elapsed) {
        this.elapsed = (0, exports.Milliseconds)(0);
        this.elapsed = elapsed;
    }
    static fromMilliseconds(milliseconds) {
        return new TimeSpan(milliseconds);
    }
    static fromSeconds(seconds) {
        return new TimeSpan(exports.Seconds.toMilliseconds(seconds));
    }
    static fromMinutes(minutes) {
        return new TimeSpan(exports.Seconds.toMilliseconds(exports.Minutes.toSeconds(minutes)));
    }
    static fromHours(hours) {
        return new TimeSpan(exports.Seconds.toMilliseconds(exports.Minutes.toSeconds(exports.Hours.toMinutes(hours))));
    }
    getTotalDays() {
        return exports.Hours.toDays(this.getTotalHours());
    }
    getTotalHours() {
        return exports.Minutes.toHours(this.getTotalMinutes());
    }
    getTotalMinutes() {
        return exports.Seconds.toMinutes(this.getTotalSeconds());
    }
    getTotalSeconds() {
        return exports.Milliseconds.toSeconds(this.elapsed);
    }
    getTotalMilliseconds() {
        return this.elapsed;
    }
    getDays() {
        return (0, exports.Days)(Math.floor(this.getTotalDays()));
    }
    getHours() {
        return (0, exports.Hours)(Math.floor(this.getTotalHours() - this.getDays() * 24));
    }
    getMinutes() {
        return (0, exports.Minutes)(Math.floor(this.getTotalMinutes() - Math.floor(this.getTotalHours()) * 60));
    }
    getSeconds() {
        return (0, exports.Seconds)(Math.floor(this.getTotalSeconds() - Math.floor(this.getTotalMinutes()) * 60));
    }
    getMilliseconds() {
        return (0, exports.Milliseconds)(this.elapsed - Math.floor(this.getTotalSeconds()) * 1000);
    }
    toString() {
        let result = this.getTotalSeconds() < 0 ? '-P' : 'P';
        let time = '';
        if (this.getHours() !== 0)
            time += `${Math.abs(this.getHours()).toString()}H`;
        if (this.getMinutes() !== 0)
            time += `${Math.abs(this.getMinutes()).toString()}M`;
        if (this.getSeconds() !== 0)
            time += `${Math.abs(this.getSeconds()).toString()}S`;
        if (this.getDays() !== 0 || time === '')
            result += `${Math.abs(this.getDays()).toString()}D`;
        if (time !== '')
            result += `T${time}`;
        return result;
    }
}
exports.TimeSpan = TimeSpan;
