// @cardano-sdk/core/dist/cjs/Serialization/TransactionBody/ProposalProcedure/HardForkInitiationAction.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _HardForkInitiationAction_protocolVersion, _HardForkInitiationAction_govActionId, _HardForkInitiationAction_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HardForkInitiationAction = void 0;
const CBOR_1 = require("../../CBOR");
const GovernanceActionId_1 = require("../../Common/GovernanceActionId");
const GovernanceActionKind_1 = require("./GovernanceActionKind");
const Governance_1 = require("../../../Cardano/types/Governance");
const util_1 = require("@cardano-sdk/util");
const Common_1 = require("../../Common");
const misc_1 = require("../../../util/misc");
const EMBEDDED_GROUP_SIZE = 3;
class HardForkInitiationAction {
    constructor(protocolVersion, govActionId) {
        _HardForkInitiationAction_protocolVersion.set(this, void 0);
        _HardForkInitiationAction_govActionId.set(this, void 0);
        _HardForkInitiationAction_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _HardForkInitiationAction_protocolVersion, protocolVersion, "f");
        __classPrivateFieldSet(this, _HardForkInitiationAction_govActionId, govActionId, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _HardForkInitiationAction_originalBytes, "f"))
            return __classPrivateFieldGet(this, _HardForkInitiationAction_originalBytes, "f");
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(GovernanceActionKind_1.GovernanceActionKind.HardForkInitiation);
        __classPrivateFieldGet(this, _HardForkInitiationAction_govActionId, "f") ? writer.writeEncodedValue((0, misc_1.hexToBytes)(__classPrivateFieldGet(this, _HardForkInitiationAction_govActionId, "f").toCbor())) : writer.writeNull();
        writer.writeEncodedValue((0, misc_1.hexToBytes)(__classPrivateFieldGet(this, _HardForkInitiationAction_protocolVersion, "f").toCbor()));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const kind = Number(reader.readUInt());
        if (kind !== GovernanceActionKind_1.GovernanceActionKind.HardForkInitiation)
            throw new util_1.InvalidArgumentError('cbor', `Expected action kind, expected ${GovernanceActionKind_1.GovernanceActionKind.HardForkInitiation} but got ${kind}`);
        let govActionId;
        if (reader.peekState() === CBOR_1.CborReaderState.Null) {
            reader.readNull();
        }
        else {
            govActionId = GovernanceActionId_1.GovernanceActionId.fromCbor(util_1.HexBlob.fromBytes(reader.readEncodedValue()));
        }
        const protocolVersion = Common_1.ProtocolVersion.fromCbor(util_1.HexBlob.fromBytes(reader.readEncodedValue()));
        const action = new HardForkInitiationAction(protocolVersion, govActionId);
        __classPrivateFieldSet(action, _HardForkInitiationAction_originalBytes, cbor, "f");
        return action;
    }
    toCore() {
        return {
            __typename: Governance_1.GovernanceActionType.hard_fork_initiation_action,
            governanceActionId: __classPrivateFieldGet(this, _HardForkInitiationAction_govActionId, "f") ? __classPrivateFieldGet(this, _HardForkInitiationAction_govActionId, "f").toCore() : null,
            protocolVersion: __classPrivateFieldGet(this, _HardForkInitiationAction_protocolVersion, "f").toCore()
        };
    }
    static fromCore(hardForkInitiationAction) {
        return new HardForkInitiationAction(Common_1.ProtocolVersion.fromCore(hardForkInitiationAction.protocolVersion), hardForkInitiationAction.governanceActionId !== null
            ? GovernanceActionId_1.GovernanceActionId.fromCore(hardForkInitiationAction.governanceActionId)
            : undefined);
    }
    govActionId() {
        return __classPrivateFieldGet(this, _HardForkInitiationAction_govActionId, "f");
    }
    protocolVersion() {
        return __classPrivateFieldGet(this, _HardForkInitiationAction_protocolVersion, "f");
    }
}
exports.HardForkInitiationAction = HardForkInitiationAction;
_HardForkInitiationAction_protocolVersion = new WeakMap(), _HardForkInitiationAction_govActionId = new WeakMap(), _HardForkInitiationAction_originalBytes = new WeakMap();
