// @cardano-sdk/core/dist/cjs/Serialization/TransactionBody/ProposalProcedure/InfoAction.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _InfoAction_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoAction = void 0;
const CBOR_1 = require("../../CBOR");
const GovernanceActionKind_1 = require("./GovernanceActionKind");
const Governance_1 = require("../../../Cardano/types/Governance");
const util_1 = require("@cardano-sdk/util");
const EMBEDDED_GROUP_SIZE = 1;
class InfoAction {
    constructor() {
        _InfoAction_originalBytes.set(this, undefined);
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _InfoAction_originalBytes, "f"))
            return __classPrivateFieldGet(this, _InfoAction_originalBytes, "f");
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(GovernanceActionKind_1.GovernanceActionKind.Info);
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const kind = Number(reader.readUInt());
        if (kind !== GovernanceActionKind_1.GovernanceActionKind.Info)
            throw new util_1.InvalidArgumentError('cbor', `Expected action kind, expected ${GovernanceActionKind_1.GovernanceActionKind.Info} but got ${kind}`);
        const action = new InfoAction();
        __classPrivateFieldSet(action, _InfoAction_originalBytes, cbor, "f");
        return action;
    }
    toCore() {
        return {
            __typename: Governance_1.GovernanceActionType.info_action
        };
    }
    static fromCore(_) {
        return new InfoAction();
    }
}
exports.InfoAction = InfoAction;
_InfoAction_originalBytes = new WeakMap();
