// @cardano-sdk/core/dist/cjs/Asset/cip67.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetNameLabelNum = exports.AssetNameLabel = void 0;
const Asset_1 = require("../Cardano/types/Asset");
const util_1 = require("@cardano-sdk/util");
const crc8_1 = __importDefault(require("./crc8"));
const ASSET_LABEL_LENGTH = 8;
const ASSET_LABEL_BRACKET = '0';
const AssetNameLabel = (value) => value;
exports.AssetNameLabel = AssetNameLabel;
exports.AssetNameLabelNum = {
    ReferenceNFT: 100,
    UserFT: 333,
    UserNFT: 222,
    UserRFT: 444,
    VirtualHandle: 0
};
const assertLabelNumInterval = (labelNum) => {
    if (labelNum < 0 || labelNum > 65535) {
        throw new util_1.InvalidArgumentError('Label num', `Label ${labelNum} out of range 0 - 65535.`);
    }
};
const checksum = (labelNumHex) => (0, crc8_1.default)(Uint8Array.from(Buffer.from(labelNumHex, 'hex')))
    .toString(16)
    .padStart(2, '0');
const isInvalidChecksum = (labelNumHex, labelChecksum) => labelChecksum !== checksum(labelNumHex);
const isInvalidLength = (label) => label.length !== ASSET_LABEL_LENGTH;
const isInvalidBracket = (lead, end) => lead !== ASSET_LABEL_BRACKET || end !== ASSET_LABEL_BRACKET;
const assetNameLabelHexToNum = (label) => {
    const labelLeadBracket = label[0];
    const labelEndBracket = label[7];
    const labelNumHex = label.slice(1, 5);
    const labelChecksum = label.slice(5, 7);
    if (isInvalidLength(label) ||
        isInvalidBracket(labelLeadBracket, labelEndBracket) ||
        isInvalidChecksum(labelNumHex, labelChecksum)) {
        return null;
    }
    return (0, exports.AssetNameLabel)(Number.parseInt(labelNumHex, 16));
};
exports.AssetNameLabel.decode = (assetName) => {
    const assetNameLabel = assetName.slice(0, ASSET_LABEL_LENGTH);
    const assetNameContent = assetName.slice(ASSET_LABEL_LENGTH);
    const assetNameLabelNum = assetNameLabelHexToNum(assetNameLabel);
    if (assetNameLabelNum === null) {
        return null;
    }
    return {
        content: (0, Asset_1.AssetName)(assetNameContent),
        label: assetNameLabelNum
    };
};
exports.AssetNameLabel.encode = (assetName, labelNum) => {
    assertLabelNumInterval(labelNum);
    const labelNumHex = labelNum.toString(16).padStart(4, ASSET_LABEL_BRACKET);
    return (0, Asset_1.AssetName)(`${ASSET_LABEL_BRACKET}${labelNumHex}${checksum(labelNumHex)}${ASSET_LABEL_BRACKET}${assetName}`);
};
