// @cardano-ogmios/client/dist/ChainSynchronization/nextBlock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNextBlockResponse = exports.handler = exports.nextBlock = void 0;
const Connection_1 = require("../Connection");
const util_1 = require("../util");
function nextBlock(socket, options) {
    return socket.send(util_1.safeJSON.stringify({
        ...Connection_1.baseRequest,
        method: 'nextBlock',
        id: options?.id
    }));
}
exports.nextBlock = nextBlock;
async function handler(response, messageHandlers, cb) {
    if (isNextBlockResponse(response)) {
        switch (response.result.direction) {
            case 'backward':
                return await messageHandlers.rollBackward({
                    point: response.result.point,
                    tip: response.result.tip
                }, cb);
            case 'forward':
                return await messageHandlers.rollForward({
                    block: response.result.block,
                    tip: response.result.tip
                }, cb);
            default:
                break;
        }
    }
}
exports.handler = handler;
function isNextBlockResponse(response) {
    return typeof response?.result?.direction !== 'undefined';
}
exports.isNextBlockResponse = isNextBlockResponse;
