// @cardano-sdk/core/dist/cjs/Serialization/AuxiliaryData/TransactionMetadata/MetadatumMap.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _MetadatumMap_map, _MetadatumMap_useIndefiniteEncoding;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadatumMap = void 0;
const CBOR_1 = require("../../CBOR");
const util_1 = require("@cardano-sdk/util");
const MetadatumList_1 = require("./MetadatumList");
const TransactionMetadatum_1 = require("./TransactionMetadatum");
const misc_1 = require("../../../util/misc");
class MetadatumMap {
    constructor() {
        _MetadatumMap_map.set(this, new Map());
        _MetadatumMap_useIndefiniteEncoding.set(this, false);
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _MetadatumMap_useIndefiniteEncoding, "f")) {
            writer.writeStartMap();
        }
        else {
            writer.writeStartMap(__classPrivateFieldGet(this, _MetadatumMap_map, "f").size);
        }
        for (const [key, value] of __classPrivateFieldGet(this, _MetadatumMap_map, "f").entries()) {
            writer.writeEncodedValue((0, misc_1.hexToBytes)(key.toCbor()));
            writer.writeEncodedValue((0, misc_1.hexToBytes)(value.toCbor()));
        }
        if (__classPrivateFieldGet(this, _MetadatumMap_useIndefiniteEncoding, "f"))
            writer.writeEndMap();
        return util_1.HexBlob.fromBytes(writer.encode());
    }
    static fromCbor(cbor) {
        const map = new MetadatumMap();
        const reader = new CBOR_1.CborReader(cbor);
        const size = reader.readStartMap();
        if (size === null)
            __classPrivateFieldSet(map, _MetadatumMap_useIndefiniteEncoding, true, "f");
        while (reader.peekState() !== CBOR_1.CborReaderState.EndMap) {
            const key = TransactionMetadatum_1.TransactionMetadatum.fromCbor((0, misc_1.bytesToHex)(reader.readEncodedValue()));
            const value = TransactionMetadatum_1.TransactionMetadatum.fromCbor((0, misc_1.bytesToHex)(reader.readEncodedValue()));
            map.insert(key, value);
        }
        reader.readEndMap();
        return map;
    }
    getLength() {
        return __classPrivateFieldGet(this, _MetadatumMap_map, "f").size;
    }
    insert(key, value) {
        __classPrivateFieldGet(this, _MetadatumMap_map, "f").set(key, value);
    }
    get(key) {
        if (!__classPrivateFieldGet(this, _MetadatumMap_map, "f"))
            return undefined;
        const element = [...__classPrivateFieldGet(this, _MetadatumMap_map, "f").entries()].find((entry) => entry[0].equals(key));
        if (!element)
            return undefined;
        return element[1];
    }
    getKeys() {
        const list = new MetadatumList_1.MetadatumList();
        for (const elem of __classPrivateFieldGet(this, _MetadatumMap_map, "f").keys()) {
            list.add(elem);
        }
        return list;
    }
    equals(other) {
        if (__classPrivateFieldGet(this, _MetadatumMap_useIndefiniteEncoding, "f") !== __classPrivateFieldGet(other, _MetadatumMap_useIndefiniteEncoding, "f"))
            return false;
        if (__classPrivateFieldGet(this, _MetadatumMap_map, "f").size !== __classPrivateFieldGet(other, _MetadatumMap_map, "f").size)
            return false;
        const thisEntries = [...__classPrivateFieldGet(this, _MetadatumMap_map, "f").entries()];
        const otherEntries = [...__classPrivateFieldGet(other, _MetadatumMap_map, "f").entries()];
        for (let i = 0; i < __classPrivateFieldGet(this, _MetadatumMap_map, "f").size; ++i) {
            if (!thisEntries[i][0].equals(otherEntries[i][0]))
                return false;
            if (!thisEntries[i][1].equals(otherEntries[i][1]))
                return false;
        }
        return true;
    }
}
exports.MetadatumMap = MetadatumMap;
_MetadatumMap_map = new WeakMap(), _MetadatumMap_useIndefiniteEncoding = new WeakMap();
