// @cardano-sdk/core/dist/cjs/Serialization/TransactionBody/VotingProcedures/Voter.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Voter_kind, _Voter_credential, _Voter_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Voter = void 0;
const CBOR_1 = require("../../CBOR");
const Address_1 = require("../../../Cardano/Address");
const util_1 = require("@cardano-sdk/util");
const VoterKind_1 = require("./VoterKind");
const Governance_1 = require("../../../Cardano/types/Governance");
const EMBEDDED_GROUP_SIZE = 2;
class Voter {
    constructor(kind, credential) {
        _Voter_kind.set(this, void 0);
        _Voter_credential.set(this, void 0);
        _Voter_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _Voter_kind, kind, "f");
        __classPrivateFieldSet(this, _Voter_credential, credential, "f");
    }
    toCbor() {
        if (__classPrivateFieldGet(this, _Voter_originalBytes, "f"))
            return __classPrivateFieldGet(this, _Voter_originalBytes, "f");
        const writer = new CBOR_1.CborWriter();
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(__classPrivateFieldGet(this, _Voter_kind, "f"));
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _Voter_credential, "f").hash, 'hex'));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        let credential;
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of ${EMBEDDED_GROUP_SIZE} elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        const hash = util_1.HexBlob.fromBytes(reader.readByteString());
        switch (kind) {
            case VoterKind_1.VoterKind.ConstitutionalCommitteeKeyHash:
            case VoterKind_1.VoterKind.DrepKeyHash:
            case VoterKind_1.VoterKind.StakePoolKeyHash:
                credential = { hash, type: Address_1.CredentialType.KeyHash };
                break;
            case VoterKind_1.VoterKind.ConstitutionalCommitteeScriptHash:
            case VoterKind_1.VoterKind.DRepScriptHash:
                credential = { hash, type: Address_1.CredentialType.ScriptHash };
                break;
            default:
                throw new util_1.InvalidStateError(`Unexpected kind value: ${kind}`);
        }
        const voter = new Voter(kind, credential);
        __classPrivateFieldSet(voter, _Voter_originalBytes, cbor, "f");
        return voter;
    }
    toCore() {
        switch (__classPrivateFieldGet(this, _Voter_kind, "f")) {
            case VoterKind_1.VoterKind.ConstitutionalCommitteeKeyHash:
                return {
                    __typename: Governance_1.VoterType.ccHotKeyHash,
                    credential: {
                        hash: __classPrivateFieldGet(this, _Voter_credential, "f").hash,
                        type: Address_1.CredentialType.KeyHash
                    }
                };
            case VoterKind_1.VoterKind.ConstitutionalCommitteeScriptHash:
                return {
                    __typename: Governance_1.VoterType.ccHotScriptHash,
                    credential: {
                        hash: __classPrivateFieldGet(this, _Voter_credential, "f").hash,
                        type: Address_1.CredentialType.ScriptHash
                    }
                };
            case VoterKind_1.VoterKind.DrepKeyHash:
                return {
                    __typename: Governance_1.VoterType.dRepKeyHash,
                    credential: {
                        hash: __classPrivateFieldGet(this, _Voter_credential, "f").hash,
                        type: Address_1.CredentialType.KeyHash
                    }
                };
            case VoterKind_1.VoterKind.DRepScriptHash:
                return {
                    __typename: Governance_1.VoterType.dRepScriptHash,
                    credential: {
                        hash: __classPrivateFieldGet(this, _Voter_credential, "f").hash,
                        type: Address_1.CredentialType.ScriptHash
                    }
                };
            case VoterKind_1.VoterKind.StakePoolKeyHash:
                return {
                    __typename: Governance_1.VoterType.stakePoolKeyHash,
                    credential: {
                        hash: __classPrivateFieldGet(this, _Voter_credential, "f").hash,
                        type: Address_1.CredentialType.KeyHash
                    }
                };
            default:
                throw new util_1.InvalidStateError(`Unexpected kind value: ${__classPrivateFieldGet(this, _Voter_kind, "f")}`);
        }
    }
    static fromCore(coreVoter) {
        let voter;
        switch (coreVoter.__typename) {
            case Governance_1.VoterType.ccHotKeyHash:
            case Governance_1.VoterType.ccHotScriptHash:
                voter = Voter.newConstitutionalCommitteeHotKey(coreVoter.credential);
                break;
            case Governance_1.VoterType.dRepKeyHash:
            case Governance_1.VoterType.dRepScriptHash:
                voter = Voter.newDrep(coreVoter.credential);
                break;
            case Governance_1.VoterType.stakePoolKeyHash:
                voter = Voter.newStakingPool(coreVoter.credential.hash);
                break;
            default:
                throw new util_1.InvalidStateError('Unexpected Voter type');
        }
        return voter;
    }
    static newConstitutionalCommitteeHotKey(credential) {
        const kind = credential.type === Address_1.CredentialType.KeyHash
            ? VoterKind_1.VoterKind.ConstitutionalCommitteeKeyHash
            : VoterKind_1.VoterKind.ConstitutionalCommitteeScriptHash;
        return new Voter(kind, credential);
    }
    static newDrep(credential) {
        const kind = credential.type === Address_1.CredentialType.KeyHash ? VoterKind_1.VoterKind.DrepKeyHash : VoterKind_1.VoterKind.DRepScriptHash;
        return new Voter(kind, credential);
    }
    static newStakingPool(keyHash) {
        return new Voter(VoterKind_1.VoterKind.StakePoolKeyHash, {
            hash: keyHash,
            type: Address_1.CredentialType.KeyHash
        });
    }
    kind() {
        return __classPrivateFieldGet(this, _Voter_kind, "f");
    }
    toConstitutionalCommitteeHotCred() {
        if (__classPrivateFieldGet(this, _Voter_kind, "f") === VoterKind_1.VoterKind.ConstitutionalCommitteeKeyHash ||
            __classPrivateFieldGet(this, _Voter_kind, "f") === VoterKind_1.VoterKind.ConstitutionalCommitteeScriptHash)
            return __classPrivateFieldGet(this, _Voter_credential, "f");
        return undefined;
    }
    toDrepCred() {
        if (__classPrivateFieldGet(this, _Voter_kind, "f") === VoterKind_1.VoterKind.DrepKeyHash || __classPrivateFieldGet(this, _Voter_kind, "f") === VoterKind_1.VoterKind.DRepScriptHash)
            return __classPrivateFieldGet(this, _Voter_credential, "f");
        return undefined;
    }
    toStakingPoolKeyHash() {
        if (__classPrivateFieldGet(this, _Voter_kind, "f") === VoterKind_1.VoterKind.StakePoolKeyHash)
            return __classPrivateFieldGet(this, _Voter_credential, "f").hash;
        return undefined;
    }
    equals(other) {
        return (__classPrivateFieldGet(this, _Voter_kind, "f") === __classPrivateFieldGet(other, _Voter_kind, "f") &&
            __classPrivateFieldGet(this, _Voter_credential, "f").type === __classPrivateFieldGet(other, _Voter_credential, "f").type &&
            __classPrivateFieldGet(this, _Voter_credential, "f").hash === __classPrivateFieldGet(other, _Voter_credential, "f").hash);
    }
}
exports.Voter = Voter;
_Voter_kind = new WeakMap(), _Voter_credential = new WeakMap(), _Voter_originalBytes = new WeakMap();
