// @cardano-ogmios/client/dist/LedgerStateQuery/query/genesisConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isQueryNetworkGenesisConfigurationConway = exports.isQueryNetworkGenesisConfigurationAlonzo = exports.isQueryNetworkGenesisConfigurationShelley = exports.isQueryNetworkGenesisConfigurationByron = exports.genesisConfiguration = void 0;
const Connection_1 = require("../../Connection");
function genesisConfiguration(context, era) {
    return (0, Connection_1.Method)({
        method: 'queryNetwork/genesisConfiguration',
        params: { era }
    }, {
        handler(response, resolve, reject) {
            if (era === 'byron' && isQueryNetworkGenesisConfigurationByron(response)) {
                resolve(response.result);
            }
            else if (era === 'shelley' && isQueryNetworkGenesisConfigurationShelley(response)) {
                resolve(response.result);
            }
            else if (era === 'alonzo' && isQueryNetworkGenesisConfigurationAlonzo(response)) {
                resolve(response.result);
            }
            else if (era === 'conway' && isQueryNetworkGenesisConfigurationConway(response)) {
                resolve(response.result);
            }
            else {
                reject(response);
            }
        }
    }, context);
}
exports.genesisConfiguration = genesisConfiguration;
function isQueryNetworkGenesisConfigurationByron(response) {
    return response?.result?.era === 'byron';
}
exports.isQueryNetworkGenesisConfigurationByron = isQueryNetworkGenesisConfigurationByron;
function isQueryNetworkGenesisConfigurationShelley(response) {
    return response?.result?.era === 'shelley';
}
exports.isQueryNetworkGenesisConfigurationShelley = isQueryNetworkGenesisConfigurationShelley;
function isQueryNetworkGenesisConfigurationAlonzo(response) {
    return response?.result?.era === 'alonzo';
}
exports.isQueryNetworkGenesisConfigurationAlonzo = isQueryNetworkGenesisConfigurationAlonzo;
function isQueryNetworkGenesisConfigurationConway(response) {
    return response?.result?.era === 'conway';
}
exports.isQueryNetworkGenesisConfigurationConway = isQueryNetworkGenesisConfigurationConway;
