// @cardano-sdk/core/dist/cjs/Serialization/Scripts/PlutusScript/PlutusV2Script.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _PlutusV2Script_compiledByteCode, _PlutusV2Script_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlutusV2Script = void 0;
const Crypto = __importStar(require("@cardano-sdk/crypto"));
const CBOR_1 = require("../../CBOR");
const util_1 = require("@cardano-sdk/util");
const Script_1 = require("../../../Cardano/types/Script");
const HASH_LENGTH_IN_BYTES = 28;
class PlutusV2Script {
    constructor(compiledByteCode) {
        _PlutusV2Script_compiledByteCode.set(this, void 0);
        _PlutusV2Script_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _PlutusV2Script_compiledByteCode, compiledByteCode, "f");
    }
    toCbor() {
        if (__classPrivateFieldGet(this, _PlutusV2Script_originalBytes, "f"))
            return __classPrivateFieldGet(this, _PlutusV2Script_originalBytes, "f");
        const writer = new CBOR_1.CborWriter();
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _PlutusV2Script_compiledByteCode, "f"), 'hex'));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const bytes = reader.readByteString();
        const script = new PlutusV2Script(util_1.HexBlob.fromBytes(bytes));
        __classPrivateFieldSet(script, _PlutusV2Script_originalBytes, cbor, "f");
        return script;
    }
    toCore() {
        return {
            __type: Script_1.ScriptType.Plutus,
            bytes: this.rawBytes(),
            version: Script_1.PlutusLanguageVersion.V2
        };
    }
    static fromCore(plutusScript) {
        if (plutusScript.version !== Script_1.PlutusLanguageVersion.V2)
            throw new util_1.InvalidArgumentError('script', 'Wrong plutus language version.');
        return new PlutusV2Script(plutusScript.bytes);
    }
    hash() {
        const bytes = `02${this.rawBytes()}`;
        const hash = Crypto.blake2b(HASH_LENGTH_IN_BYTES).update(Buffer.from(bytes, 'hex')).digest();
        return Crypto.Hash28ByteBase16(util_1.HexBlob.fromBytes(hash));
    }
    rawBytes() {
        return __classPrivateFieldGet(this, _PlutusV2Script_compiledByteCode, "f");
    }
    setRawBytes(bytes) {
        __classPrivateFieldSet(this, _PlutusV2Script_compiledByteCode, bytes, "f");
        __classPrivateFieldSet(this, _PlutusV2Script_originalBytes, undefined, "f");
    }
}
exports.PlutusV2Script = PlutusV2Script;
_PlutusV2Script_compiledByteCode = new WeakMap(), _PlutusV2Script_originalBytes = new WeakMap();
