// @cardano-sdk/crypto/dist/esm/Ed25519e/Ed25519Signature.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Ed25519Signature_signature;
import { Ed25519SignatureHex } from '../hexTypes.js';
import { InvalidArgumentError } from '@cardano-sdk/util';
export const ED25519_SIGNATURE_LENGTH = 64;
export class Ed25519Signature {
    constructor(signature) {
        _Ed25519Signature_signature.set(this, void 0);
        __classPrivateFieldSet(this, _Ed25519Signature_signature, signature, "f");
    }
    static fromBytes(signature) {
        if (signature.length !== ED25519_SIGNATURE_LENGTH)
            throw new InvalidArgumentError('signature', `signature should be ${ED25519_SIGNATURE_LENGTH} bytes; however ${signature.length} bytes were provided.`);
        return new Ed25519Signature(signature);
    }
    static fromHex(signature) {
        return Ed25519Signature.fromBytes(Buffer.from(signature, 'hex'));
    }
    bytes() {
        return __classPrivateFieldGet(this, _Ed25519Signature_signature, "f");
    }
    hex() {
        return Ed25519SignatureHex(Buffer.from(__classPrivateFieldGet(this, _Ed25519Signature_signature, "f")).toString('hex'));
    }
}
_Ed25519Signature_signature = new WeakMap();
