// @cardano-sdk/crypto/dist/esm/Ed25519e/Ed25519PublicKey.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Ed25519PublicKey_keyMaterial;
import { ED25519_PUBLIC_KEY_HASH_LENGTH, Ed25519KeyHash } from './Ed25519KeyHash.js';
import { Ed25519PublicKeyHex } from '../hexTypes.js';
import { InvalidArgumentError } from '@cardano-sdk/util';
import sodium from 'libsodium-wrappers-sumo';
export const ED25519_PUBLIC_KEY_LENGTH = 32;
export class Ed25519PublicKey {
    constructor(keyMaterial) {
        _Ed25519PublicKey_keyMaterial.set(this, void 0);
        __classPrivateFieldSet(this, _Ed25519PublicKey_keyMaterial, keyMaterial, "f");
    }
    async verify(signature, message) {
        await sodium.ready;
        return sodium.crypto_sign_verify_detached(signature.bytes(), Buffer.from(message, 'hex'), __classPrivateFieldGet(this, _Ed25519PublicKey_keyMaterial, "f"));
    }
    static fromBytes(keyMaterial) {
        if (keyMaterial.length !== ED25519_PUBLIC_KEY_LENGTH)
            throw new InvalidArgumentError('keyMaterial', `Key should be ${ED25519_PUBLIC_KEY_LENGTH} bytes; however ${keyMaterial.length} bytes were provided.`);
        return new Ed25519PublicKey(keyMaterial);
    }
    static fromHex(keyMaterial) {
        return Ed25519PublicKey.fromBytes(Buffer.from(keyMaterial, 'hex'));
    }
    async hash() {
        await sodium.ready;
        const hash = sodium.crypto_generichash(ED25519_PUBLIC_KEY_HASH_LENGTH, __classPrivateFieldGet(this, _Ed25519PublicKey_keyMaterial, "f"));
        return Ed25519KeyHash.fromBytes(hash);
    }
    bytes() {
        return __classPrivateFieldGet(this, _Ed25519PublicKey_keyMaterial, "f");
    }
    hex() {
        return Ed25519PublicKeyHex(Buffer.from(__classPrivateFieldGet(this, _Ed25519PublicKey_keyMaterial, "f")).toString('hex'));
    }
}
_Ed25519PublicKey_keyMaterial = new WeakMap();
