// @cardano-sdk/core/dist/cjs/Cardano/ChainId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainIds = exports.NetworkMagics = exports.NetworkId = void 0;
var NetworkId;
(function (NetworkId) {
    NetworkId[NetworkId["Mainnet"] = 1] = "Mainnet";
    NetworkId[NetworkId["Testnet"] = 0] = "Testnet";
})(NetworkId = exports.NetworkId || (exports.NetworkId = {}));
var NetworkMagics;
(function (NetworkMagics) {
    NetworkMagics[NetworkMagics["Mainnet"] = 764824073] = "Mainnet";
    NetworkMagics[NetworkMagics["Preprod"] = 1] = "Preprod";
    NetworkMagics[NetworkMagics["Preview"] = 2] = "Preview";
    NetworkMagics[NetworkMagics["Sanchonet"] = 4] = "Sanchonet";
})(NetworkMagics = exports.NetworkMagics || (exports.NetworkMagics = {}));
exports.ChainIds = {
    Mainnet: {
        networkId: NetworkId.Mainnet,
        networkMagic: NetworkMagics.Mainnet
    },
    Preprod: {
        networkId: NetworkId.Testnet,
        networkMagic: NetworkMagics.Preprod
    },
    Preview: {
        networkId: NetworkId.Testnet,
        networkMagic: NetworkMagics.Preview
    },
    Sanchonet: {
        networkId: NetworkId.Testnet,
        networkMagic: NetworkMagics.Sanchonet
    }
};
