// @cardano-sdk/core/dist/cjs/Serialization/Scripts/NativeScript/ScriptPubkey.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ScriptPubkey_keyHash, _ScriptPubkey_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScriptPubkey = void 0;
const Crypto = __importStar(require("@cardano-sdk/crypto"));
const CBOR_1 = require("../../CBOR");
const util_1 = require("@cardano-sdk/util");
const Script_1 = require("../../../Cardano/types/Script");
const EMBEDDED_GROUP_SIZE = 2;
class ScriptPubkey {
    constructor(keyHash) {
        _ScriptPubkey_keyHash.set(this, void 0);
        _ScriptPubkey_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _ScriptPubkey_keyHash, keyHash, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _ScriptPubkey_originalBytes, "f"))
            return __classPrivateFieldGet(this, _ScriptPubkey_originalBytes, "f");
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(Script_1.NativeScriptKind.RequireSignature);
        writer.writeByteString(Buffer.from(__classPrivateFieldGet(this, _ScriptPubkey_keyHash, "f"), 'hex'));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of two elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== Script_1.NativeScriptKind.RequireSignature)
            throw new util_1.InvalidArgumentError('cbor', `Expected kind ${Script_1.NativeScriptKind.RequireSignature}, but got kind ${kind}`);
        const key = Crypto.Ed25519KeyHashHex(util_1.HexBlob.fromBytes(reader.readByteString()));
        const script = new ScriptPubkey(key);
        __classPrivateFieldSet(script, _ScriptPubkey_originalBytes, cbor, "f");
        return script;
    }
    toCore() {
        return {
            __type: Script_1.ScriptType.Native,
            keyHash: __classPrivateFieldGet(this, _ScriptPubkey_keyHash, "f"),
            kind: Script_1.NativeScriptKind.RequireSignature
        };
    }
    static fromCore(script) {
        return new ScriptPubkey(script.keyHash);
    }
    keyHash() {
        return __classPrivateFieldGet(this, _ScriptPubkey_keyHash, "f");
    }
    setKeyHash(keyHash) {
        __classPrivateFieldSet(this, _ScriptPubkey_keyHash, keyHash, "f");
        __classPrivateFieldSet(this, _ScriptPubkey_originalBytes, undefined, "f");
    }
}
exports.ScriptPubkey = ScriptPubkey;
_ScriptPubkey_keyHash = new WeakMap(), _ScriptPubkey_originalBytes = new WeakMap();
