// @cardano-sdk/core/dist/cjs/Serialization/Scripts/NativeScript/ScriptNOfK.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ScriptNOfK_nativeScripts, _ScriptNOfK_required, _ScriptNOfK_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScriptNOfK = void 0;
const CBOR_1 = require("../../CBOR");
const util_1 = require("@cardano-sdk/util");
const NativeScript_1 = require("./NativeScript");
const Script_1 = require("../../../Cardano/types/Script");
const EMBEDDED_GROUP_SIZE = 3;
class ScriptNOfK {
    constructor(nativeScripts, required) {
        _ScriptNOfK_nativeScripts.set(this, void 0);
        _ScriptNOfK_required.set(this, void 0);
        _ScriptNOfK_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _ScriptNOfK_nativeScripts, nativeScripts, "f");
        __classPrivateFieldSet(this, _ScriptNOfK_required, required, "f");
    }
    toCbor() {
        const writer = new CBOR_1.CborWriter();
        if (__classPrivateFieldGet(this, _ScriptNOfK_originalBytes, "f"))
            return __classPrivateFieldGet(this, _ScriptNOfK_originalBytes, "f");
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(Script_1.NativeScriptKind.RequireNOf);
        writer.writeInt(__classPrivateFieldGet(this, _ScriptNOfK_required, "f"));
        writer.writeStartArray(__classPrivateFieldGet(this, _ScriptNOfK_nativeScripts, "f").length);
        for (const nativeScript of __classPrivateFieldGet(this, _ScriptNOfK_nativeScripts, "f"))
            writer.writeEncodedValue(Buffer.from(nativeScript.toCbor(), 'hex'));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of two elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== Script_1.NativeScriptKind.RequireNOf)
            throw new util_1.InvalidArgumentError('cbor', `Expected kind ${Script_1.NativeScriptKind.RequireNOf}, but got kind ${kind}`);
        const required = reader.readInt();
        const scripts = new Array();
        reader.readStartArray();
        while (reader.peekState() !== CBOR_1.CborReaderState.EndArray)
            scripts.push(NativeScript_1.NativeScript.fromCbor(util_1.HexBlob.fromBytes(reader.readEncodedValue())));
        reader.readEndArray();
        const script = new ScriptNOfK(scripts, Number(required));
        __classPrivateFieldSet(script, _ScriptNOfK_originalBytes, cbor, "f");
        return script;
    }
    toCore() {
        return {
            __type: Script_1.ScriptType.Native,
            kind: Script_1.NativeScriptKind.RequireNOf,
            required: __classPrivateFieldGet(this, _ScriptNOfK_required, "f"),
            scripts: __classPrivateFieldGet(this, _ScriptNOfK_nativeScripts, "f").map((script) => script.toCore())
        };
    }
    static fromCore(script) {
        return new ScriptNOfK(script.scripts.map((nativeScript) => NativeScript_1.NativeScript.fromCore(nativeScript)), script.required);
    }
    required() {
        return __classPrivateFieldGet(this, _ScriptNOfK_required, "f");
    }
    setRequired(required) {
        __classPrivateFieldSet(this, _ScriptNOfK_required, required, "f");
        __classPrivateFieldSet(this, _ScriptNOfK_originalBytes, undefined, "f");
    }
    nativeScripts() {
        return __classPrivateFieldGet(this, _ScriptNOfK_nativeScripts, "f");
    }
    setNativeScripts(nativeScripts) {
        __classPrivateFieldSet(this, _ScriptNOfK_nativeScripts, nativeScripts, "f");
        __classPrivateFieldSet(this, _ScriptNOfK_originalBytes, undefined, "f");
    }
}
exports.ScriptNOfK = ScriptNOfK;
_ScriptNOfK_nativeScripts = new WeakMap(), _ScriptNOfK_required = new WeakMap(), _ScriptNOfK_originalBytes = new WeakMap();
