// @cardano-sdk/core/dist/cjs/Cardano/types/DelegationsAndRewards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cip17FromMetadatum = exports.portfolioMetadataFromCip17 = exports.DelegationMetadataLabel = exports.StakeCredentialStatus = void 0;
const metadatum_1 = require("../../util/metadatum");
var StakeCredentialStatus;
(function (StakeCredentialStatus) {
    StakeCredentialStatus["Registering"] = "REGISTERING";
    StakeCredentialStatus["Registered"] = "REGISTERED";
    StakeCredentialStatus["Unregistering"] = "UNREGISTERING";
    StakeCredentialStatus["Unregistered"] = "UNREGISTERED";
})(StakeCredentialStatus = exports.StakeCredentialStatus || (exports.StakeCredentialStatus = {}));
exports.DelegationMetadataLabel = 6862n;
const portfolioMetadataFromCip17 = (cip17) => {
    const portfolio = { ...cip17 };
    portfolio.pools = cip17.pools.map((pool) => ({
        id: pool.id,
        weight: pool.weight
    }));
    return portfolio;
};
exports.portfolioMetadataFromCip17 = portfolioMetadataFromCip17;
const cip17FromMetadatum = (portfolio) => {
    const cip17 = (0, metadatum_1.metadatumToJson)(portfolio);
    for (const pool of cip17.pools) {
        pool.weight = Number(pool.weight);
    }
    return cip17;
};
exports.cip17FromMetadatum = cip17FromMetadatum;
