// @cardano-sdk/core/dist/cjs/Cardano/types/Governance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDRepAlwaysNoConfidence = exports.isDRepAlwaysAbstain = exports.isDRepCredential = exports.VoterType = exports.Vote = exports.GovernanceActionType = void 0;
var GovernanceActionType;
(function (GovernanceActionType) {
    GovernanceActionType["parameter_change_action"] = "parameter_change_action";
    GovernanceActionType["hard_fork_initiation_action"] = "hard_fork_initiation_action";
    GovernanceActionType["treasury_withdrawals_action"] = "treasury_withdrawals_action";
    GovernanceActionType["no_confidence"] = "no_confidence";
    GovernanceActionType["update_committee"] = "update_committee";
    GovernanceActionType["new_constitution"] = "new_constitution";
    GovernanceActionType["info_action"] = "info_action";
})(GovernanceActionType = exports.GovernanceActionType || (exports.GovernanceActionType = {}));
var Vote;
(function (Vote) {
    Vote[Vote["no"] = 0] = "no";
    Vote[Vote["yes"] = 1] = "yes";
    Vote[Vote["abstain"] = 2] = "abstain";
})(Vote = exports.Vote || (exports.Vote = {}));
var VoterType;
(function (VoterType) {
    VoterType["ccHotKeyHash"] = "ccHotKeyHash";
    VoterType["ccHotScriptHash"] = "ccHotScriptHash";
    VoterType["dRepKeyHash"] = "dRepKeyHash";
    VoterType["dRepScriptHash"] = "dRepScriptHash";
    VoterType["stakePoolKeyHash"] = "stakePoolKeyHash";
})(VoterType = exports.VoterType || (exports.VoterType = {}));
const isDRepCredential = (deleg) => !('__typename' in deleg);
exports.isDRepCredential = isDRepCredential;
const isDRepAlwaysAbstain = (deleg) => !(0, exports.isDRepCredential)(deleg) && deleg.__typename === 'AlwaysAbstain';
exports.isDRepAlwaysAbstain = isDRepAlwaysAbstain;
const isDRepAlwaysNoConfidence = (deleg) => !(0, exports.isDRepCredential)(deleg) && deleg.__typename === 'AlwaysNoConfidence';
exports.isDRepAlwaysNoConfidence = isDRepAlwaysNoConfidence;
