// @cardano-sdk/core/dist/cjs/Cardano/types/Script.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPlutusScript = exports.isNativeScript = exports.PlutusLanguageVersion = exports.NativeScriptKind = exports.ScriptType = void 0;
var ScriptType;
(function (ScriptType) {
    ScriptType["Native"] = "native";
    ScriptType["Plutus"] = "plutus";
})(ScriptType = exports.ScriptType || (exports.ScriptType = {}));
var NativeScriptKind;
(function (NativeScriptKind) {
    NativeScriptKind[NativeScriptKind["RequireSignature"] = 0] = "RequireSignature";
    NativeScriptKind[NativeScriptKind["RequireAllOf"] = 1] = "RequireAllOf";
    NativeScriptKind[NativeScriptKind["RequireAnyOf"] = 2] = "RequireAnyOf";
    NativeScriptKind[NativeScriptKind["RequireNOf"] = 3] = "RequireNOf";
    NativeScriptKind[NativeScriptKind["RequireTimeAfter"] = 4] = "RequireTimeAfter";
    NativeScriptKind[NativeScriptKind["RequireTimeBefore"] = 5] = "RequireTimeBefore";
})(NativeScriptKind = exports.NativeScriptKind || (exports.NativeScriptKind = {}));
var PlutusLanguageVersion;
(function (PlutusLanguageVersion) {
    PlutusLanguageVersion[PlutusLanguageVersion["V1"] = 0] = "V1";
    PlutusLanguageVersion[PlutusLanguageVersion["V2"] = 1] = "V2";
    PlutusLanguageVersion[PlutusLanguageVersion["V3"] = 2] = "V3";
})(PlutusLanguageVersion = exports.PlutusLanguageVersion || (exports.PlutusLanguageVersion = {}));
const isNativeScript = (script) => script.__type === ScriptType.Native;
exports.isNativeScript = isNativeScript;
const isPlutusScript = (script) => script.__type === ScriptType.Plutus;
exports.isPlutusScript = isPlutusScript;
